pkg/aws/appmesh.go

```diff
@@ -220,65 +220,80 @@ func (v *VirtualNode) Listeners() []appmeshv1beta1.Listener {
-				TimeoutMillis:      appmeshListener.HealthCheck.TimeoutMillis,
-				UnhealthyThreshold: appmeshListener.HealthCheck.UnhealthyThreshold,
+				HealthyThreshold:   sdkListener.HealthCheck.HealthyThreshold,
+				IntervalMillis:     sdkListener.HealthCheck.IntervalMillis,
+				Path:               sdkListener.HealthCheck.Path,
+				Port:               sdkListener.HealthCheck.Port,
+				Protocol:           sdkListener.HealthCheck.Protocol,
+				TimeoutMillis:      sdkListener.HealthCheck.TimeoutMillis,
+				UnhealthyThreshold: sdkListener.HealthCheck.UnhealthyThreshold,
 			}
 			listener.HealthCheck = healthCheck
 		}
+		if sdkListener.Tls != nil {
+			cert := appmeshv1beta1.ListenerTlsCertificate{}
+			if sdkListener.Tls.Certificate.Acm != nil {
+				cert.Acm = &appmeshv1beta1.ListenerTlsAcmCertificate{
+					CertificateArn: aws.StringValue(sdkListener.Tls.Certificate.Acm.CertificateArn),
+				}
+			}
+			if sdkListener.Tls.Certificate.File != nil {
+				cert.File = &appmeshv1beta1.ListenerTlsFileCertificate{
+					CertificateChain: aws.StringValue(sdkListener.Tls.Certificate.File.CertificateChain),
+					PrivateKey:       aws.StringValue(sdkListener.Tls.Certificate.File.PrivateKey),
+				}
+			}
+			listener.Tls = &appmeshv1beta1.ListenerTls{
+				Mode:        aws.StringValue(sdkListener.Tls.Mode),
+				Certificate: cert,
+			}
+		}
 		listeners = append(listeners, listener)
 	}
 	return listeners
 }
 
-// ListenersSet converts into a Set of Listeners
-func (v *VirtualNode) ListenersSet() set.Set {
-	listeners := v.Listeners()
-	s := set.NewSet()
-	for i := range listeners {
-		s.Add(listeners[i])
-	}
-	return s
-}
-
 // Backends converts into our API type
 func (v *VirtualNode) Backends() []appmeshv1beta1.Backend {
 	if v.Data.Spec.Backends == nil {
 		return []appmeshv1beta1.Backend{}
 	}
 
 	var backends = []appmeshv1beta1.Backend{}
-	for _, b := range v.Data.Spec.Backends {
-		backends = append(backends, appmeshv1beta1.Backend{
+	for _, sdkBackend := range v.Data.Spec.Backends {
+		crdBackend := appmeshv1beta1.Backend{
 			VirtualService: appmeshv1beta1.VirtualServiceBackend{
-				VirtualServiceName: aws.StringValue(b.VirtualService.VirtualServiceName),
+				VirtualServiceName: aws.StringValue(sdkBackend.VirtualService.VirtualServiceName),
 			},
-		})
+		}
+		if sdkBackend.VirtualService.ClientPolicy != nil {
+			crdBackend.VirtualService.ClientPolicy = convertSdkClientPolicyToCrd(sdkBackend.VirtualService.ClientPolicy)
+		}
+		backends = append(backends, crdBackend)
 	}
 	return backends
 }
 
-// BackendsSet returns a set of Backends defined for virtual-node
-func (v *VirtualNode) BackendsSet() set.Set {
-	backends := v.Backends()
-	s := set.NewSet()
-	for i := range backends {
-		s.Add(backends[i])
+// BackendDefaults converts from the SDK types into CRD types
+func (v *VirtualNode) BackendDefaults() *appmeshv1beta1.BackendDefaults {
+	if v.Data.Spec.BackendDefaults == nil {
+		return &appmeshv1beta1.BackendDefaults{}
```

Comment:
nit, what's the semantic empty value for BackendDefaults? is it always `&appmeshv1beta1.BackendDefaults{}` or `nil`? currently it's fine since we will always do null-checks for BackendDefaults.ClientPolicy(since it's a pointer). But these unclearly defined empty value is typical a source a bugs.

Reply:
Good catch. This should be nil. It's an optional field in the API.

Reply:
I see why I did it, the other helpers like `Listeners()` and `Backends()` define it as an empty object/slice. Possibly due to API semantics.